File: workers/go/cmd/http-worker/render.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// render expands {{var:NAME}} and {{task_id}} placeholders in template.
// Mirrors workers/rust/crates/http-worker/src/render.rs render().
//
// Unknown placeholders (anything other than var:NAME and task_id) are left
// in place so the operator notices via the rendered request.
func render(template string, vars map[string]any, taskID string) string {
	var out strings.Builder
	out.Grow(len(template))
	rest := template
	for {
		i := strings.Index(rest, "{{")
		if i < 0 {
			out.WriteString(rest)
			return out.String()
		}
		out.WriteString(rest[:i])
		afterOpen := rest[i+2:]
		j := strings.Index(afterOpen, "}}")
		if j < 0 {
			out.WriteString("{{")
			rest = afterOpen
			continue
		}
		key := strings.TrimSpace(afterOpen[:j])
		out.WriteString(resolve(key, vars, taskID))
		rest = afterOpen[j+2:]
	}
}

func resolve(key string, vars map[string]any, taskID string) string {
	if key == "task_id" {
		return taskID
	}
	if name, ok := strings.CutPrefix(key, "var:"); ok {
		v, present := vars[name]
		if !present {
			return ""
		}
		return scalarString(v)
	}
	// Unknown placeholder — leave intact.
	return "{{" + key + "}}"
}

func scalarString(v any) string {
	switch t := v.(type) {
	case nil:
		return ""
	case string:
		return t
	case bool:
		return strconv.FormatBool(t)
	case float64:
		// JSON numbers decode as float64; keep integer-valued ones integer-formatted.
		if t == float64(int64(t)) {
			return strconv.FormatInt(int64(t), 10)
		}
		return strconv.FormatFloat(t, 'f', -1, 64)
	case int, int32, int64:
		return fmt.Sprintf("%d", t)
	default:
		return fmt.Sprintf("%v", t)
	}
}

// renderJSON walks template recursively, expanding placeholders inside
// strings. Strings that are exactly `{{var:NAME}}` for a non-string variable
// retain the underlying type via passthrough — matches the Rust SDK's
// single_var_passthrough behaviour.
func renderJSON(template any, vars map[string]any, taskID string) any {
	switch t := template.(type) {
	case string:
		if v, ok := singleVarPassthrough(t, vars); ok {
			return v
		}
		return render(t, vars, taskID)
	case []any:
		out := make([]any, len(t))
		for i, v := range t {
			out[i] = renderJSON(v, vars, taskID)
		}
		return out
	case map[string]any:
		out := make(map[string]any, len(t))
		for k, v := range t {
			out[k] = renderJSON(v, vars, taskID)
		}
		return out
	default:
		return t
	}
}

// singleVarPassthrough: if s is exactly `{{var:NAME}}` (with optional
// whitespace) and NAME is in vars, return the underlying value.
func singleVarPassthrough(s string, vars map[string]any) (any, bool) {
	trimmed := strings.TrimSpace(s)
	inner, ok := strings.CutPrefix(trimmed, "{{")
	if !ok {
		return nil, false
	}
	inner, ok = strings.CutSuffix(inner, "}}")
	if !ok {
		return nil, false
	}
	name, ok := strings.CutPrefix(strings.TrimSpace(inner), "var:")
	if !ok {
		return nil, false
	}
	v, present := vars[strings.TrimSpace(name)]
	if !present {
		return nil, false
	}
	return v, true
}

// jsonpath applies a minimal JSON-path subset to root: $.foo, $.foo.bar,
// $.foo[0]. Anything else returns ok=false so the operator notices via a
// missing variable.
func jsonpath(expr string, root any) (any, bool) {
	expr, ok := strings.CutPrefix(expr, "$.")
	if !ok {
		return nil, false
	}
	if expr == "" {
		return root, true
	}
	current := root
	for _, segment := range strings.Split(expr, ".") {
		name := segment
		var idx *int
		if i := strings.Index(segment, "["); i >= 0 {
			name = segment[:i]
			rest, ok := strings.CutSuffix(segment[i+1:], "]")
			if !ok {
				return nil, false
			}
			n, err := strconv.Atoi(rest)
			if err != nil {
				return nil, false
			}
			idx = &n
		}
		if name != "" {
			obj, ok := current.(map[string]any)
			if !ok {
				return nil, false
			}
			current, ok = obj[name]
			if !ok {
				return nil, false
			}
		}
		if idx != nil {
			arr, ok := current.([]any)
			if !ok {
				return nil, false
			}
			if *idx < 0 || *idx >= len(arr) {
				return nil, false
			}
			current = arr[*idx]
		}
	}
	return current, true
}

File: workers/crates/http-worker/src/render.rs
use serde_json::Value;
use std::collections::HashMap;

/// Tiny `{{var:NAME}}` and `{{task_id}}` substitution. We deliberately
/// stay below mustache because anything fancier (conditionals, loops)
/// belongs in the BPMN, not in the worker config.
pub fn render(template: &str, vars: &HashMap<String, Value>, task_id: &str) -> String {
    let mut out = String::with_capacity(template.len());
    let mut rest = template;
    while let Some(start) = rest.find("{{") {
        out.push_str(&rest[..start]);
        let after_open = &rest[start + 2..];
        let Some(end) = after_open.find("}}") else {
            out.push_str("{{");
            rest = after_open;
            continue;
        };
        let key = after_open[..end].trim();
        let replacement = resolve(key, vars, task_id);
        out.push_str(&replacement);
        rest = &after_open[end + 2..];
    }
    out.push_str(rest);
    out
}

fn resolve(key: &str, vars: &HashMap<String, Value>, task_id: &str) -> String {
    if key == "task_id" {
        return task_id.to_string();
    }
    if let Some(name) = key.strip_prefix("var:") {
        return match vars.get(name) {
            Some(Value::String(s)) => s.clone(),
            Some(other) => other.to_string(),
            None => String::new(),
        };
    }
    // Unknown placeholder — leave it intact so the operator notices.
    format!("{{{{{key}}}}}")
}

/// Render a JSON template, recursively replacing strings.
pub fn render_json(template: &Value, vars: &HashMap<String, Value>, task_id: &str) -> Value {
    match template {
        Value::String(s) => {
            let rendered = render(s, vars, task_id);
            // Try to keep the type — if the rendered string is exactly a
            // variable substitution and the variable's underlying type is
            // not a string, return the original JSON value.
            if let Some(v) = single_var_passthrough(s, vars) {
                v.clone()
            } else {
                Value::String(rendered)
            }
        }
        Value::Array(arr) => {
            Value::Array(arr.iter().map(|v| render_json(v, vars, task_id)).collect())
        }
        Value::Object(obj) => Value::Object(
            obj.iter()
                .map(|(k, v)| (k.clone(), render_json(v, vars, task_id)))
                .collect(),
        ),
        other => other.clone(),
    }
}

/// If `s` is exactly `{{var:NAME}}` and `NAME` is in `vars`, return the
/// underlying JSON value so non-string types (numbers, booleans) survive
/// the round-trip.
fn single_var_passthrough<'a>(s: &str, vars: &'a HashMap<String, Value>) -> Option<&'a Value> {
    let trimmed = s.trim();
    let inner = trimmed.strip_prefix("{{")?.strip_suffix("}}")?.trim();
    let name = inner.strip_prefix("var:")?.trim();
    vars.get(name)
}

/// Apply a JSON-path-ish expression to a response body. We support a
/// minimal subset: `$.foo`, `$.foo.bar`, `$.foo[0]`. Anything else
/// returns `None` so the operator notices via the missing variable.
pub fn jsonpath<'a>(expr: &str, root: &'a Value) -> Option<&'a Value> {
    let expr = expr.strip_prefix("$.")?.trim();
    if expr.is_empty() {
        return Some(root);
    }
    let mut current = root;
    for segment in expr.split('.') {
        let (name, idx) = match segment.find('[') {
            Some(i) => {
                let name = &segment[..i];
                let rest = &segment[i + 1..];
                let idx = rest.strip_suffix(']')?.parse::<usize>().ok()?;
                (name, Some(idx))
            }
            None => (segment, None),
        };
        if !name.is_empty() {
            current = current.get(name)?;
        }
        if let Some(i) = idx {
            current = current.get(i)?;
        }
    }
    Some(current)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn vars(pairs: &[(&str, Value)]) -> HashMap<String, Value> {
        pairs
            .iter()
            .map(|(k, v)| (k.to_string(), v.clone()))
            .collect()
    }

    #[test]
    fn render_substitutes_variables() {
        let v = vars(&[("name", json!("alice"))]);
        assert_eq!(render("hi {{var:name}}!", &v, "tid"), "hi alice!");
    }

    #[test]
    fn render_substitutes_task_id() {
        let v = HashMap::new();
        assert_eq!(render("k-{{task_id}}", &v, "abc"), "k-abc");
    }

    #[test]
    fn render_unknown_placeholder_left_intact() {
        let v = HashMap::new();
        assert_eq!(render("{{var:missing}}", &v, "tid"), "");
        assert_eq!(render("{{weird}}", &v, "tid"), "{{weird}}");
    }

    #[test]
    fn render_json_preserves_number_via_passthrough() {
        let v = vars(&[("amount", json!(1000))]);
        let template = json!({ "amount": "{{var:amount}}" });
        assert_eq!(render_json(&template, &v, "tid"), json!({ "amount": 1000 }));
    }

    #[test]
    fn render_json_interpolates_inside_string() {
        let v = vars(&[("name", json!("alice"))]);
        let template = json!({ "greeting": "hi {{var:name}}" });
        assert_eq!(
            render_json(&template, &v, "tid"),
            json!({ "greeting": "hi alice" })
        );
    }

    #[test]
    fn jsonpath_simple_field() {
        let body = json!({ "id": "abc" });
        assert_eq!(jsonpath("$.id", &body), Some(&json!("abc")));
    }

    #[test]
    fn jsonpath_nested() {
        let body = json!({ "a": { "b": 42 } });
        assert_eq!(jsonpath("$.a.b", &body), Some(&json!(42)));
    }

    #[test]
    fn jsonpath_array_index() {
        let body = json!({ "items": [10, 20, 30] });
        assert_eq!(jsonpath("$.items[1]", &body), Some(&json!(20)));
    }

    #[test]
    fn jsonpath_missing_returns_none() {
        let body = json!({});
        assert_eq!(jsonpath("$.missing", &body), None);
    }
}
